spark/v3.2/spark-extensions/src/main/scala/org/apache/spark/sql/catalyst/analysis/AssignmentAlignmentSupport.scala

```diff
@@ -0,0 +1,250 @@
+import org.apache.spark.sql.catalyst.expressions.ExtractValue
+import org.apache.spark.sql.catalyst.expressions.GetStructField
+import org.apache.spark.sql.catalyst.expressions.Literal
+import org.apache.spark.sql.catalyst.expressions.NamedExpression
+import org.apache.spark.sql.catalyst.plans.logical.Assignment
+import org.apache.spark.sql.catalyst.plans.logical.LogicalPlan
+import org.apache.spark.sql.catalyst.util.CharVarcharUtils
+import org.apache.spark.sql.internal.SQLConf.StoreAssignmentPolicy
+import org.apache.spark.sql.types.DataType
+import org.apache.spark.sql.types.StructField
+import org.apache.spark.sql.types.StructType
+import scala.collection.mutable
+
+trait AssignmentAlignmentSupport extends CastSupport {
+
+  self: SQLConfHelper =>
+
+  private case class ColumnUpdate(ref: Seq[String], expr: Expression)
+
+  /**
+   * Checks whether assignments are aligned and match table columns.
+   *
+   * @param table a target table
+   * @param assignments assignments to check
+   * @return true if the assignments are aligned
+   */
+  protected def aligned(table: LogicalPlan, assignments: Seq[Assignment]): Boolean = {
+    val sameSize = table.output.size == assignments.size
+    sameSize && table.output.zip(assignments).forall { case (attr, assignment) =>
+      val key = assignment.key
+      val value = assignment.value
+      asAssignmentRef(attr) == asAssignmentRef(key) && attr.dataType == value.dataType
+    }
+  }
+
+  /**
+   * Aligns assignments to match table columns.
+   * <p>
+   * This method processes and reorders given assignments so that each target column gets
+   * an expression it should be set to. If a column does not have a matching assignment,
+   * it will be set to its current value. For example, if one passes a table with columns c1, c2
+   * and an assignment c2 = 1, this method will return c1 = c1, c2 = 1.
+   * <p>
+   * This method also handles updates to nested columns. If there is an assignment to a particular
+   * nested field, this method will construct a new struct with one field updated
+   * preserving other fields that have not been modified. For example, if one passes a table with
+   * columns c1, c2 where c2 is a struct with fields n1 and n2 and an assignment c2.n2 = 1,
+   * this method will return c1 = c1, c2 = struct(c2.n1, 1).
+   *
+   * @param table a target table
+   * @param assignments assignments to align
+   * @return aligned assignments that match table columns
+   */
+  protected def alignAssignments(
+      table: LogicalPlan,
+      assignments: Seq[Assignment]): Seq[Assignment] = {
+
+    val columnUpdates = assignments.map(a => ColumnUpdate(asAssignmentRef(a.key), a.value))
+    val outputExprs = applyUpdates(table.output, columnUpdates)
+    outputExprs.zip(table.output).map {
+      case (expr, attr) => toAssignment(attr, expr)
+    }
+  }
+
+  protected def toAssignment(key: Expression, value: Expression): Assignment = {
```

Comment:
Some comments to explain what is going on here would be really helpful. Maybe also rename this to something that makes it more clear, like `handleCharTypeLimit`.

If you end up renaming this, also consider accepting an Assignment and optionally transforming it, like this:

```scala
    protected def handleCharLimits(assignment: Assignment): Assignment = {
    val rawKeyType = ...
    if (CharVarcharUtils.hasCharVarchar(rawKeyType)) {
      // build the transformed assignment
    } else {
      assignment
    }
  }
```

Reply:
I added `AssignmentUtils` with a separate method dedicated to this functionality.